src/main/java/com/microsoft/sqlserver/jdbc/SharedTimer.java

```diff
@@ -56,16 +59,17 @@ static synchronized boolean isRunning() {
      *
      * If the reference count reaches zero then the underlying executor will be shutdown so that its thread stops.
      */
-    public synchronized void removeRef() {
-        if (refCount <= 0) {
-            throw new IllegalStateException("removeRef() called more than actual references");
-        }
-        refCount -= 1;
-        if (refCount == 0) {
-            // Removed last reference so perform cleanup
-            executor.shutdownNow();
-            executor = null;
-            instance = null;
+    public void removeRef() {
+        synchronized (lock) {
+            if (refCount.get() <= 0) {
+                throw new IllegalStateException("removeRef() called more than actual references");
```

Comment:
@cheenamalhotra , shouldn't this be a `SQLServerException` instead?

Reply:
It's not SQL Server Exception or driver generated exception if this code is reached. This is driver's internal class and the APIs are called internally. So technically if this exception is thrown, it is an illegal state of the driver/application just like a "NullPointerException" that can occur anytime a null value is referenced.

Reply:
Is the error message meant to be localized?

Reply:
This is "non-expected" error as it's driver internal implementation. It's not meant to occur.
So no, we wouldn't need to localize it. If this was a driver thrown exception, we would have localized it when this was added.